draft-ietf-webpush-encryption.md

```diff
@@ -108,170 +101,277 @@ encryption scheme by applications that use push messaging.  An agent that only
+  {{!I-D.ietf-httpbis-encryption-encoding}} with an expanded context string
+  ({{context}}).
+
+* Encryption or decryption follows according to
+  {{!I-D.ietf-httpbis-encryption-encoding}}.
+
+The key derivation process is summarized in {{summary}}.  Restrictions on the
+use of the encrypted content coding are described in {{restrict}}.
+
+
+## Diffie-Hellman Key Agreement {#dh}
+
+For each new subscription that the User Agent generates for an application, it
+also generates a P-256 {{FIPS186}} key pair for use in elliptic-curve
+Diffie-Hellman (ECDH) {{ECDH}}.
 
-The application using the subscription distributes the key identifier and public
-key along with other subscription information, such as the subscription URI and
-expiration time.
+When sending a push message, the Application Server also generates a new ECDH
+key pair on the same P-256 curve.
 
-The communication medium by which an application distributes the key identifier
-and public key MUST be confidentiality protected for the reasons described in
-[I-D.ietf-webpush-protocol].  Most applications that use push messaging have a
-pre-existing relationship with an Application Server.  Any existing
-communication mechanism that is authenticated and provides confidentiality and
-integrity, such as HTTPS [RFC2818], is sufficient.
+The ECDH public key for the Application Server is included in the `dh` parameter
+of the Crypto-Key header field (see {{iana}}).  The uncompressed point form
+defined in {{X9.62}} (that is, a 65 octet sequence that starts with a 0x04
+octet) is encoded using base64url {{!RFC7515}} to produce the `dh` parameter
+value.
+
+An application combines its ECDH private key with the public key provided by the
+User Agent using the process described in {{ECDH}}; on receipt of the push
+message, a User Agent combines its private key with the public key provided by
+the Application Server in the `dh` parameter in the same way.  These operations
+produce the same value for the ECDH shared secret.
 
 
 ## Push Message Authentication {#auth}
 
 To ensure that push messages are correctly authenticated, a symmetric
 authentication secret is added to the information generated by a User Agent.
 The authentication secret is mixed into the key derivation process described in
-[I-D.ietf-httpbis-encryption-encoding].
-
-The authentication secret ensures that exposure or leakage of the DH public
-key - which, as a public key, is not necessarily treated as a secret - does not
-enable an adversary to generate valid push messages.
+{{!I-D.ietf-httpbis-encryption-encoding}}.
 
-A User Agent MUST generate and provide a hard to guess sequence of octets that
+A User Agent MUST generate and provide a hard to guess sequence of 16 octets that
 is used for authentication of push messages.  This SHOULD be generated by a
-cryptographically strong random number generator [RFC4086] and be at least 16
-octets long.
+cryptographically strong random number generator {{!RFC4086}}.
```

Comment:
Hard to guess is the actual requirement.  You can actually generate authentication secrets by (for example) deriving them from a global secret and the subscription identifier.  That's not random, and it saves you from storing a teeny bit of state.  I realize that I'm splitting hairs, but still...